src/common/hist_util.h

```diff
@@ -196,23 +196,135 @@ size_t DeviceSketch(int device,
  *  Transform floating values to integer index in histogram This is a global histogram
  *  index for CPU histogram.  On GPU ellpack page is used.
  */
+
+
+enum BinBounds
+{
+  POWER_OF_TWO_8,
+  POWER_OF_TWO_16,
+  POWER_OF_TWO_32,
+};
+
+uint32_t getValueFromUint8(void *t, size_t i);
+/*{
+  std::cout << "\nuint8_t\n";
+  return 0;
+  //return ((uint8_t*)t)[i];
+}*/
+uint32_t getValueFromUint16(void* t, size_t i);
+/*{
+  std::cout << "\nuint32_t\n";
+  return 0;//((uint16_t*)t)[i];
+}
+*/
+uint32_t getValueFromUint32(void* t, size_t i);
+/*{
+  std::cout << "\nuint32_t\n";
+  return 0;//((uint32_t*)t)[i];
+}*/
+
+
+typedef uint32_t (*Func)(void*, size_t);
+
+struct Index
+{
+  Index(): binBound_(POWER_OF_TWO_8), p_(1), data_(nullptr), disp_(nullptr)
+  {
+    setBinBound(binBound_);
+  }
+  ~Index()
+  {
+    if(data_)
+    {
+      char* delete_ptr = static_cast<char*>(data_);
```

Comment:
some magic here :)
I suppose you can use ```std::vector<char> data_; ``` instead of row pointers

Reply:
done

Reply:
Thanks
Anyway, please, remove the comments